his_checker.py

```diff
@@ -71,6 +71,32 @@ def process_calls(files):
     print(f"== Check done with {metric_fail} error(s) ==\n")
     return metric_fail
 
+def process_recursive(files):
+
+    """Detect if there are any recursive functions."""
+
+    metric_fail = 0
+    cflow = "cflow -l -r "
+    print("== Checking the number of recursive functions ==")
+    pattern = r'\{\s*(\d+)\}\s+(.+?)\s+<(.+?)\sat\s(.+):(\d+)> \(R\)'
+    pattern_re = re.compile(pattern)
+    for file in files.files:
+        if file.endswith(".c"):
```

Comment:
This can't be right. Since you check one file at a time, this will only detect functions that directly recurse on themselves, or maybe that recurse through some other function that is in the same source file. Also you don't considered `static inline` functions that can be defined in header files.

This check is defined by HIS metrics as for checking for _"Call graph recursions"_. Therefore, the case when one function A calls a function B defined in a different source files, that itself calls function A, must be detected.

Reply:
This is a good observation. However, after some testing the `cflow` tool can observe the following recursions for this example https://github.com/bao-project/bao-his-tests/tree/main/src/ap_ag_cycle :

```
$ cflow -l -Isrc/ap_ag_cycle/inc src/ap_ag_cycle/ap_ag_cycle1.c src/ap_ag_cycle/ap_ag_cycle2.c --no-cpp
{   0} main() <int main (void) at src/ap_ag_cycle/ap_ag_cycle1.c:65>:
{   1}     funcA() <void funcA (int val) at src/ap_ag_cycle/ap_ag_cycle1.c:10> (R):
{   2}         printf()
{   2}         funcA() <void funcA (int val) at src/ap_ag_cycle/ap_ag_cycle1.c:10> (recursive: see 2)
{   1}     funcB() <void funcB (int *val) at src/ap_ag_cycle/ap_ag_cycle1.c:29> (R):
{   2}         printf()
{   2}         funcC() <void funcC (void) at src/ap_ag_cycle/ap_ag_cycle1.c:42> (R):
{   3}             printf()
{   3}             funcD() <void funcD (void) at src/ap_ag_cycle/ap_ag_cycle1.c:47> (R):
{   4}                 printf()
{   4}                 funcB() <void funcB (int *val) at src/ap_ag_cycle/ap_ag_cycle1.c:29> (recursive: see 5)
{   1}     funcE() <void funcE (int *val) at src/ap_ag_cycle/ap_ag_cycle1.c:52>:
{   2}         printf()
{   2}         funcF()
```

The `(R)` notation marks functions that recursively call themselves. However, as I see from the output that detecting the case that you described (>function A calls a function B defined in a different source files, that itself calls function A) can't be detected. From another search, I still didn't found another alternatives for a tool to this particular case. I would keep looking.